Prepared package with ADAM 6224 Device Class

Patfra opened the pull request:
Closes #3

1. Working ADAM 6224 Device Class  
- reading from Digital Inputs
- reading/writing to Analog Outputs
- reading/writing to Type Code of Analog Outputs
- reading/writing to Safety Value of Analog Outputs
- reading/writing to Startup Value  of Analog Outputs
- reading Status of Analog Output
- reading Event Status of Digital Inputs
2. Established package hierarchy 
3. Added LICENCE 

The documentation will be added later

Patfra committed:
First working version of ADAM-6224 class
- prepared Adam6224 class
- removed testing files (modbus-test.py, MyTango.py)
- updated .gitignore

Patfra committed:
Changed project structure #3
- added LICENCE
- established package hierarchy

Patfra committed:
#3
Small changes
- added reconnecting when connection is lost
- connection to device is now closed when deleting an instance

LJBD commented on adam/adam_6224_ds/__init__.py:
```diff
@@ -0,0 +1,5 @@
+"""
+Adam6224 description
```
Please provide a better description here.

LJBD commented on adam/adam_6224_ds/adam_6224.py:
```diff
@@ -0,0 +1,355 @@
+# See LICENSE.txt for more info.
+
+""" Adam6224
+It is a definicion of a class used to control ADAM-6224 controller via Modbus/TCP.
```
Typo in "definition". Also, please use CamelCase for the names - meaning that `Adam6224` should be renamed to `ADAM6224`, since that's its original spelling.

Patfra replied:
Do i have to change names of packages to ADAM62XX?

LJBD replied:
No, packages should be named with `snake_case`. Only in case of `CamelCase` you should keep the original capital letters of "ADAM" and we use `CamelCase` for class names and Tango interface elements (attributes, properties and commands).

LJBD commented on adam/adam_6224_ds/adam_6224.py:
```diff
@@ -0,0 +1,355 @@
+# PROTECTED REGION END #    //  Adam6224.additionnal_import
+
+
+class Adam6224(Device):
```
Please rename the class to `ADAM6224`.

LJBD commented on adam/adam_6224_ds/adam_6224.py:
```diff
@@ -0,0 +1,355 @@
+
+class Adam6224(Device):
+    """
+    It is a definicion of a class used to control ADAM-6224 controller via Modbus/TCP.
```
Please provide a better description here. You can write about the types and quantities of its connections.

Patfra committed:
#3 Requested changes and updates
- changed class name to ADAM6224
- added better description
- added more features to ADAM6224 class
- changed structure of packages
- added run_server.py script

LJBD commented on adam/__init__.py:
```diff
@@ -0,0 +1,26 @@
+__author__ = "Patryk Fraczek"
+__author_email__ = "pat.fraczekC@gmail.com"
+
+for package_name in find_packages():
```
This loop is unnecessary.

Patfra commented:
@LJBD, you seen changes and updates i committed yesterday? Is there anything else I should add or change?

LJBD commented on adam/adam_6224_ds/adam_6224.py:
```diff
@@ -0,0 +1,355 @@
+
```
Why are there two files named `adam_6224.py`?

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+from tango.server import class_property, device_property
+from tango import AttrQuality, AttrWriteType, DispLevel, DevState
+# Additional import
+# PROTECTED REGION ID(ADAM6224.additional_import) ENABLED START #
```
Please remove all comments for `PROTECTED REGIONS` as they won't be needed anymore.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+It is a definition of a class used to control ADAM-6224 controller via Modbus TCP.
+Device contains 4-ch Isolated Analog Output and 4-ch Digital Input.
+For each AO there are associated attributes:
+ - AnalogOutput(double) - contains present output value of the channel (V or mA), it is adjustable
```
I'm fairly certain that this list will not be rendered correctly in Sphinx.
Please surround the lists in docstrings with empty lines and begin items with an asterisk, as in:
```
The line preceding a list.

* the first element,
* the second element

The line after a list.
```

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    event_status_dictionary = {'Unreliable DI value (UART Timeout)': int(1), 'Safety Value triggered': int(2),
+                            'Startup Value triggered': int(4), ' ': int(0)}
+    event_status_dictionary.update({v: k for k, v in event_status_dictionary.iteritems()})
```
Please create a different variable for the reverse dictionary.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+    analog_output_safety_values = [0, 0, 0, 0]
+    type_code_dictionary = {'0-20mA': int("0182", 16), '4-20mA': int("0180", 16), '0-10V': int("0148", 16),
+                            '0-5V': int("0147", 16), '+-10V': int("0143", 16), '+-5V': int("0142", 16)}
+    type_code_dictionary.update({v: k for k, v in type_code_dictionary.iteritems()})
```
Please create a different variable for the reverse dictionary.

LJBD replied:
One can be named `type_to_code_dict` and the other `code_to_type_dict`.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    status_dict_1 = {'Fail to provide AO Value': int(1), 'No Output Current': int(8),
+                     'Zero/Span Calibration Error': int(512), ' ': int(0)}
+    status_dict_1.update({v: k for k, v in status_dict_1.iteritems()})
```
Please create a different variable for the reverse dictionary.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    status_dict_2 = {'DI triggered to Safety Value': int(1),'DI triggered to Startup Value': int(2),
+                     'AO triggered to Fail Safety Value': int(4), ' ': int(0)}
+    status_dict_2.update({v: k for k, v in status_dict_2.iteritems()})
```
Please create a different variable for the reverse dictionary.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+    def init_device(self):
+        Device.init_device(self)
+        # PROTECTED REGION ID(ADAM6224.init_device) ENABLED START #
+        self.set_state(DevState.ON)
```
I wouldn't use `ON` state as the default if you don't connect to an ADAM at initialisation. Please change that to `STANDBY` and update the status accordingly as well.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+        self.set_status("ADAM-6224 enabled")
+        # PROTECTED REGION END #    //  ADAM6224.init_device
+
+    def always_executed_hook(self):
```
This method is unnecessary, you can remove it completely.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    def read_DigitalInput_0(self):
+        # PROTECTED REGION ID(ADAM6224.DigitalInput_0_read) ENABLED START #
+        tmp = self.connected_ADAM.read_coils(0, 4)
```
I see that you're reading some data from ADAM device only in reading the `DigitalInput_0`. Why are you not reading it in all of the other methods as well?

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    def read_EventStatus_0(self):
+        # PROTECTED REGION ID(ADAM6224.EventStatus_0_read) ENABLED START #
+        tmp = self.connected_ADAM.read_holding_registers(110, 4)
```
Same situation here. The reading from the device is done only for the `EventStatus_0`. That doesn't look correct.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+        tmp = self.connected_ADAM.read_holding_registers(100, 8)
+        self.analog_output_statuses = tmp.registers
+        print(self.analog_output_statuses)
+        return self.status_dict_1[self.analog_output_statuses[0]] + ' ' + self.status_dict_2[self.analog_output_statuses[1]]
```
Line longer than 80 characters.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    def read_Status_1(self):
+        # PROTECTED REGION ID(ADAM6224.Status_1_read) ENABLED START #
+        return self.status_dict_1[self.analog_output_statuses[2]] + ' ' + self.status_dict_2[self.analog_output_statuses[3]]
```
Line longer than 80 characters.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    def read_Status_2(self):
+        # PROTECTED REGION ID(ADAM6224.Status_2_read) ENABLED START #
+        return self.status_dict_1[self.analog_output_statuses[4]] + ' ' + self.status_dict_2[self.analog_output_statuses[5]]
```
Line longer than 80 characters.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    def read_Status_3(self):
+        # PROTECTED REGION ID(ADAM6224.Status_3_read) ENABLED START #
+        return self.status_dict_1[self.analog_output_statuses[6]] + ' ' + self.status_dict_2[self.analog_output_statuses[7]]
```
Line longer than 80 characters.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+        except Exception as e:
+            self.set_state(DevState.FAULT)
+            self.set_status("Exception caught while connecting to device:\n%s" % e)
+        print("Connected do device with IP: " + self.DeviceAddress)
```
This should not be printed but rather displayed as a status or a log message.

LJBD replied:
In general, please don't use `print` in Device Classes. Use logging methods like: `debug_stream`, `info_stream`, `warning_stream` and `error_stream`.

LJBD commented on adam/run_server.py:
```diff
@@ -0,0 +1,18 @@
+
+
+
+"""
```
This docstring is empty.

LJBD commented on adam/adam_6224.py:
```diff
@@ -0,0 +1,628 @@
+
+    # --------
+    # Commands
+    # --------
```
I'd suggest to do all of the reading from ADAMs in a dedicated, polled command. It should update variables of the Tango Device. That way an operation of reading an attribute doesn't affect the controlled device.

Patfra replied:
Can this command looks something like that:
```python
def read_DataFromDevice(self):
        #read Digital Inputs
        tmp = self.connected_ADAM.read_coils(0, 4)
        self.digital_input_values = tmp.bits
        # read Analog Outputs Value
        tmp = self.connected_ADAM.read_holding_registers(0, 4)
        self.analog_output_values = tmp.registers
        # read Analog Outputs Status
        tmp = self.connected_ADAM.read_holding_registers(100, 8)
        self.analog_output_statuses = tmp.registers
        # read Digital Inputs Status
        tmp = self.connected_ADAM.read_holding_registers(110, 4)
        self.digital_input_events = tmp.registers
        # read Type Codes
        tmp = self.connected_ADAM.read_holding_registers(200, 4)
        self.analog_output_types = tmp.registers
        # read Startup Values
        tmp = self.connected_ADAM.read_holding_registers(400, 4)
        self.analog_output_startup_values = tmp.registers
        # read Safety Values
        tmp = self.connected_ADAM.read_holding_registers(410, 4)
        self.analog_output_safety_values = tmp.registers'

LJBD replied:
Yes, that's the idea. Remember that you can specify the default polling period in your `@command` decorator and that you shouldn't try to read when you're not connected to a physical device. I'd suggest checking if the connection is made to the beginning of this method.

Patfra committed:
#3 Minor changes
- deleted duplicate of adam_6224.py
- deleted PROTECTED REGION comments
- fixed docstrings
- fixed dictionaries names
- fixed lines length in adam_6224.py
- all reading methods put in single function
- added description to attributes
